src/main/java/io/strimzi/kafka/quotas/StaticQuotaCallback.java

```diff
@@ -145,49 +121,47 @@ public boolean updateClusterMetadata(Cluster cluster) {
     public void close() {
         try {
             closeExecutorService();
+            volumeSourceBuilder.close();
         } finally {
-            Metrics.defaultRegistry().allMetrics().keySet().stream().filter(m -> scope.equals(m.getScope())).forEach(Metrics.defaultRegistry()::removeMetric);
+            Metrics.defaultRegistry().allMetrics().keySet().stream().filter(m -> SCOPE.equals(m.getScope())).forEach(Metrics.defaultRegistry()::removeMetric);
         }
     }
 
+    @SuppressWarnings("deprecation")
     @Override
     public void configure(Map<String, ?> configs) {
         StaticQuotaConfig config = new StaticQuotaConfig(configs, true);
         quotaMap = config.getQuotaMap();
-        storageQuotaSoft = config.getSoftStorageQuota();
-        storageQuotaHard = config.getHardStorageQuota();
+        final Optional<Long> availableBytesLimitConfig = config.getAvailableBytesLimit();
+
+        if (availableBytesLimitConfig.isPresent()) {
+            throttleFactorSupplier = new AvailableBytesThrottleFactorSupplier(availableBytesLimitConfig.get());
+            log.info("Available bytes limit {}", availableBytesLimitConfig.get());
+        } else {
+            storageQuotaSoft = config.getSoftStorageQuota();
+            storageQuotaHard = config.getHardStorageQuota();
+
+            throttleFactorSupplier = new TotalConsumedThrottleFactorSupplier(storageQuotaHard, storageQuotaSoft);
+            log.info("Total consumed Storage quota (soft, hard): ({}, {}).", storageQuotaSoft, storageQuotaHard);
+        }
+
+        throttleFactorSupplier.addUpdateListener(() -> resetQuota.add(ClientQuotaType.PRODUCE));
         excludedPrincipalNameList = config.getExcludedPrincipalNameList();
 
         long storageCheckIntervalMillis = TimeUnit.SECONDS.toMillis(config.getStorageCheckInterval());
 
         if (storageCheckIntervalMillis > 0L) {
-            List<Path> logDirs = config.getLogDirs().stream().map(Paths::get).collect(Collectors.toList());
-            storageChecker.configure(
-                    logDirs,
-                    this::updateUsedStorage);
-            backgroundScheduler.scheduleWithFixedDelay(storageChecker, storageCheckIntervalMillis, storageCheckIntervalMillis, TimeUnit.MILLISECONDS);
-            log.info("Configured quota callback with {}. Storage quota (soft, hard): ({}, {}). Storage check interval: {}ms", quotaMap, storageQuotaSoft, storageQuotaHard, storageCheckIntervalMillis);
+            Runnable volumeSource = volumeSourceBuilder.withConfig(config).withVolumeConsumer(throttleFactorSupplier).build();
+            backgroundScheduler.scheduleWithFixedDelay(volumeSource, 0, storageCheckIntervalMillis, TimeUnit.MILLISECONDS);
```

Comment:
Although you've scheduled with fixed delay, the code that runs is actually only submitting requests to the `Admin` client. Those requests will be completed on the `Admin` client's completion thread. So if some of the requests you make using the Admin client take a long time to complete (e.g. requests start timing out) then it's possible the scheduler just keeps adding more and more requests to the `Admin`'s request queue.

Was that intentional? If so, it deserves Javadoc about why. If not, it might be simplest to just make `ClusterVolumeSource.run()` block by using `get` on the futures.

Reply:
Resolved with https://github.com/strimzi/kafka-quotas-plugin/pull/33/commits/139b97146eb3e6e61bdaefdce13998fa922a5986